cli/src/cmd_net.rs

```diff
@@ -0,0 +1,1485 @@
+            }
+        }
+
+        Ok(())
+    }
+}
+
+/// Get the status of BGP.
+#[derive(Parser, Debug, Clone)]
+#[command(verbatim_doc_comment)]
+#[command(name = "net bgp status")]
+pub struct CmdBgpStatus {}
+
+#[async_trait]
+impl RunnableCmd for CmdBgpStatus {
+    async fn run(&self, ctx: &Context) -> Result<()> {
+        let c = ctx.client()?;
+
+        let status = c.networking_bgp_status().send().await?.into_inner();
+
+        let (sw0, sw1) = status
+            .iter()
+            .partition(|x| x.switch == SwitchLocation::Switch0);
+
+        println!("{}", "switch0".dimmed());
+        println!("{}", "=======".dimmed());
+        show_status(&sw0)?;
+        println!();
+
+        println!("{}", "switch1".dimmed());
+        println!("{}", "=======".dimmed());
+        show_status(&sw1)?;
+
+        Ok(())
+    }
+}
+
+fn show_status(st: &Vec<&BgpPeerStatus>) -> Result<()> {
+    let mut tw = TabWriter::new(std::io::stdout()).ansi(true);
+    writeln!(
+        &mut tw,
+        "{}\t{}\t{}\t{}\t{}",
+        "Peer Address".dimmed(),
+        "Local ASN".dimmed(),
+        "Remote ASN".dimmed(),
+        "Session State".dimmed(),
+        "State Duration".dimmed(),
+    )?;
+    for s in st {
+        writeln!(
+            tw,
+            "{}\t{}\t{}\t{:?}\t{}",
+            s.addr,
+            s.local_asn,
+            s.remote_asn,
+            s.state,
+            humantime::Duration::from(std::time::Duration::from_millis(s.state_duration_millis)),
+        )?;
+    }
+    tw.flush()?;
+    Ok(())
+}
+
+/// Get the status of switch ports.
+#[derive(Parser, Debug, Clone)]
+#[command(verbatim_doc_comment)]
+#[command(name = "net port status")]
+pub struct CmdPortStatus {}
+
+#[async_trait]
+impl RunnableCmd for CmdPortStatus {
+    async fn run(&self, ctx: &Context) -> Result<()> {
+        let c = ctx.client()?;
+
+        let ports = c
+            .networking_switch_port_list()
+            .limit(u32::MAX)
+            .send()
+            .await?
+            .into_inner()
+            .items;
+
+        let (mut sw0, mut sw1): (Vec<&SwitchPort>, Vec<&SwitchPort>) = ports
+            .iter()
+            .partition(|x| x.switch_location.as_str() == "switch0");
+
+        sw0.sort_by_key(|x| x.port_name.as_str());
+        sw1.sort_by_key(|x| x.port_name.as_str());
+
+        println!("{}", "switch0".dimmed());
+        println!("{}", "=======".dimmed());
+        self.show_switch(c, "switch0", &sw0).await?;
+
+        println!("{}", "switch1".dimmed());
+        println!("{}", "=======".dimmed());
+        self.show_switch(c, "switch1", &sw1).await?;
+
+        Ok(())
+    }
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug)]
+pub struct MacAddr {
+    a: [u8; 6],
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug)]
+struct LinkStatus {
+    address: MacAddr,
+    enabled: bool,
+    autoneg: bool,
+    fec: String,
+    link_state: String,
+    fsm_state: String,
+    media: String,
+    speed: String,
+}
+
```

Comment:
how fragile is this?

Reply:
Quite. The link status endpoint is experimental and opaque, passing just `Object` back. We could, in theory, import types from Dendrite, which is where these types ultimately come from, but that repo is not open, so there are challenges on that front.